Create a singular INTERNAL.TASK_LOG table which records query_id

joshelser opened the pull request:
Suggestion from Jacques on Friday after we discussed the slow performance of #643 (due to the task_history and query_history UDTFs).

* Separate what a task does from the metadata recording that we want to track
* Record the query_id for a task when it runs into that metadata
* Migrate the separate task log tables into one (with even context to filter the update of a specific object)

joshelser commented on bootstrap/090_post_setup.sql:
```diff
@@ -70,21 +70,84 @@ CREATE OR REPLACE TASK TASKS.WAREHOUSE_EVENTS_MAINTENANCE
+DECLARE
+    task_name text default 'SIMPLE_DATA_EVENTS_MAINTENANCE';
+    task_run_id text default (select INTERNAL.TASK_RUN_ID());
+    query_id text default (select query_id from table(information_schema.task_history(TASK_NAME => 'SIMPLE_DATA_EVENTS_MAINTENANCE')) WHERE GRAPH_RUN_GROUP_ID = :task_run_id  AND DATABASE_NAME = current_database() limit 1);
```
I'd love this to be a UDF; however, because this uses a context_function, it has to live in a task.

We could make a task DAG that ensures a task which creates a common UDF is run before any other task, but this felt overkill for the size of this PR already.

rymurr replied:
I have a slight preference for this (and the other similar calls) to live in the `BEGIN`/`END` block. My general rule of thumb was to put only very simple statements into the `DECLAR` block. You do similar on eg line 142

joshelser replied:
fixed (and swapped the string literal for the `:task_name` variable -- wasn't sure if that was going to work inside the UDTF argument list)

rymurr commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,54 @@
+            FROM INTERNAL.TASK_WAREHOUSE_EVENTS twe;
+    end if;
+
+    if (exists (select 1 from information_schema.tables where table_schema = 'INTERNAL' and table_name = 'TASK_SIMPLE_DATA_EVENTS')) then
```
not we use `1` here and `*` everywhere else

rymurr commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,54 @@
+AS
+BEGIN
+    -- If the old task log tables exist, copy the rows into the new task_log table
+    if (exists (select * from information_schema.tables where table_schema = 'INTERNAL' and table_name = 'TASK_QUERY_HISTORY')) then
```
It looks like we will copy data from the old to the new if the old exists. Where do we delete the old and/or prevent multiple runnings of this creating dub data

joshelser replied:
Oh, duh. I was thinking about the "next" upgrade and didn't think about the "next-next" upgrade.

joshelser replied:
will truncate old log tables after insert.

rymurr commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,54 @@
+     SYSTEM$TASK_RUNTIME_INFO('CURRENT_TASK_GRAPH_RUN_GROUP_ID')
+$$;
+
+let task_log_exists boolean := (exists (select * from information_schema.tables where table_schema = 'INTERNAL' and table_name = 'TASK_LOG'));
```
nit: creating the varialbe here an using it so much later in the script was slightly jarring to me

joshelser replied:
since truncating the old tables after inserting the rows into the new task_log table, this boolean is just an optimization and not necessary for correctness. Dropped it.

joshelser commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+$$;
+
+-- Generic table that tasks should record their execution into.
+CREATE TABLE INTERNAL.TASK_LOG IF NOT EXISTS (task_start timestamp_ltz, success boolean, task_name varchar, object_name varchar,
+    input variant, output variant, task_finish timestamp_ltz, task_run_id text, query_id text);
```
I hadn't included the min/max range of data materialized (e.g. we have query history data from 2023-01-01 00:00:00 to 2023-04-30 05:30:00). because I thought this would be easy to dynamically compute after the fact.

However, it was not as simple as I expected to dynamically construct and run the `select min(..), max(...) from ..` queries. I'm leaning towards to including two new columns (range_min, range_max) into the task_log table.

rymurr replied:
I thoght the min and max were included in the output variant?

joshelser replied:
updated `output` so it includes `range_min` and `range_max`

rymurr commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+BEGIN
+    -- If the old task log tables exist, copy the rows into the new task_log table
+    if (exists (select * from information_schema.tables where table_schema = 'INTERNAL' and table_name = 'TASK_QUERY_HISTORY')) then
+        INSERT INTO INTERNAL.TASK_LOG (task_start, success, task_name, object_name, input, output, task_finish)
```
can you put these ( and the below) into a transaction? The insert and teh trunat. Just to ensure they happen atomically

joshelser replied:
Ugh, now Github shows me your comment.

I ended up switching the truncate table .. to a drop table (we don't need to keep these empty tables laying around). I think your ask is then moot as the DDL is going to auto-commit. LMK if I missed something.

joshelser replied:
changed again after Jacques' feedback to not be destructive over the old task log tables.

This sets a value in internal.config to indicate we've migrated the task logs to the new table.

joshelser commented:
I can't find your comment in Github (hidden now because I pushed a change that affected the line, I think)

> can you put these ( and the below) into a transaction? The insert and teh trunat. Just to ensure they happen atomically

I ended up switching the `truncate table ..` to a `drop table` (we don't need to keep these empty tables laying around). I think your ask is then moot as the DDL is going to auto-commit. LMK if I missed something.

jacques-n commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+$$;
+
+-- Generic table that tasks should record their execution into.
+CREATE TABLE INTERNAL.TASK_LOG IF NOT EXISTS (task_start timestamp_ltz, success boolean, task_name varchar, object_name varchar,
```
should we have oldest and newest data fields in this? I'd rather use structure anywhere we can.

jacques-n commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+        INSERT INTO INTERNAL.TASK_LOG (task_start, success, task_name, object_name, input, output, task_finish)
+            SELECT tqh.run, tqh.success, 'QUERY_HISTORY_MAINTENANCE', 'QUERY_HISTORY', tqh.input, tqh.output, null
+            FROM INTERNAL.TASK_QUERY_HISTORY tqh;
+        DROP TABLE INTERNAL.TASK_QUERY_HISTORY;
```
I don't think we should do this. In this version. I think we should do in future version. If we have a problem and roll back, I think the app would think that a person didn't do full materialization. This would at minimum cause a bunch of extra compute and more likely also cause wrong result (i think). This applies to all the drops/truncates.

joshelser replied:
> If we have a problem and roll back, I think the app would think that a person didn't do full materialization. This would at minimum cause a bunch of extra compute and more likely also cause wrong result (i think).

Yeah, you're exactly right -- it would try to re-materialize things. I'll come up with a different solution that can preserve the old tables.

jacques-n commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+    end if;
+
+    if (exists (select * from information_schema.tables where table_schema = 'INTERNAL' and table_name = 'TASK_WAREHOUSE_EVENTS')) then
+        INSERT INTO INTERNAL.TASK_LOG (task_start, success, task_name, object_name, input, output, task_finish)
```
Does this code work on a system that doesn't have this table? I'm thinking that you aren't using dynamic sql and I'm not confident on behavior of parsing/if statement behavior in this case.

joshelser replied:
I thought so, but I'll double check.

jacques-n commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+call internal.migrate_task_logs();
+
+-- Create specific-purpose views that we had before.
+CREATE OR REPLACE VIEW REPORTING.SIMPLE_DATA_EVENTS_TASK_HISTORY AS SELECT * exclude (task_name) rename (object_name as table_name)
```
Was this a view before? I'm not convinced that the name is good.

joshelser replied:
Yeah, these three (adjacent in this file) views existed prior.

joshelser replied:
With the introduction of `REPORTING.TASK_LOG_HISTORY` (user-facing view for the internal.task_log table), we could just drop these old user-facing views. They are of no value for users of the native app -- just for developers.

jacques-n commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,53 @@
+-- Create specific-purpose views that we had before.
+CREATE OR REPLACE VIEW REPORTING.SIMPLE_DATA_EVENTS_TASK_HISTORY AS SELECT * exclude (task_name) rename (object_name as table_name)
+    FROM INTERNAL.TASK_LOG WHERE task_name = 'SIMPLE_DATA_EVENTS_MAINTENANCE';
+CREATE OR REPLACE VIEW REPORTING.WAREHOUSE_LOAD_EVENTS_TASK_HISTORY AS SELECT * exclude (task_name) rename (object_name as warehouse_name)
```
would be good to also structure input/output variants for each of the types if we're creating these domain-specific views.

joshelser replied:
Resolving as I dropped views. If we want to reintroduce any of these domain-specific views, I think it would be good to be intentional on the schema.

jacques-n commented on bootstrap/018_warehouse_updates.sql:
```diff
@@ -34,7 +33,7 @@ begin
 end;
 
-CREATE OR REPLACE PROCEDURE internal.refresh_warehouse_events(migrate boolean) RETURNS STRING LANGUAGE SQL
+CREATE OR REPLACE PROCEDURE internal.refresh_warehouse_events(migrate boolean, input variant) RETURNS OBJECT LANGUAGE SQL
```
given that variant is input, we should declare (via comment or similar) what the expected fields must be/could be.

joshelser replied:
> given that variant is input

The previous code was doing cast to `VARIANT` for no apparent reason. I unwound some of that (after I missed updating a `RETURNS`). I'll just swap them all to be OBJECT since that's what is intended.

jacques-n replied:
I'm speaking to the fields in the object/variant not the type itself. 

It used to be that the object was entirely inside the procedure. You've moved it to the signature which is why I'm suggesting we document the expected structure.

joshelser replied:
Ah, I see. Makes sense

jacques-n commented on bootstrap/018_warehouse_updates.sql:
```diff
@@ -82,9 +80,22 @@ BEGIN
         ELSE
-            insert into INTERNAL.TASK_WAREHOUSE_EVENTS SELECT :dt, true, :input, OBJECT_CONSTRUCT('oldest_running', :oldest_running, 'newest_completed', :newest_completed, 'attempted_migrate', :migrate, 'migrate', :migrate1, 'migrate_INCOMPLETE', :migrate2, 'new_records', 0, 'new_INCOMPLETE', 0, 'new_closed', 0)::VARIANT;
+            let range_min timestamp := (select min(session_end) from (
+                select min(session_end) as session_end from INTERNAL_REPORTING_MV.CLUSTER_AND_WAREHOUSE_SESSIONS_COMPLETE_AND_DAILY
```
why hand constructing the union as opposed to just using existing object?

joshelser replied:
The higher-level views which union the complete/incomplete tables also include a filter on `session_type = 'CL'` or `session_type = 'WH'`.

I reckon it would be more correct to deduce the exact range of data available for both cluster and warehouse sessions separately. But, that makes the warehouse materialization case different from the rest. Feasible, but more complex.

jacques-n replied:
Remember that we should be having a row in the user-facing table for each view that we maintain. The user-facing data should be focused on internal tables/views. So in the case of clsuter session and warehouse session, there should be at least two rows (i think more since there is daily and hourly versions I believe)

joshelser replied:
> there should be at least two rows (i think more since there is daily and hourly versions I believe)

Yeah, I thought I could get away with a single task_log row for one execution of the warehouse_events_maintenance task, but the distinction between cluster_sessions and warehouse_sessions means I have to change this logic.

The daily and hourly variants are easily included when we're actually creating the materialization status view since they are (usually) backed by the same internal table. Clusters/sessions are the exception. I'll break this out.

jacques-n commented on bootstrap/019_query_history_updates.sql:
```diff
@@ -67,20 +63,32 @@ BEGIN
-            insert into INTERNAL.TASK_QUERY_HISTORY SELECT :run_id, true, :input, OBJECT_CONSTRUCT('oldest_running', :oldest_running, 'newest_completed', :newest_completed, 'attempted_migrate', :migrate, 'migrate', :migrate1, 'migrate_INCOMPLETE', :migrate2, 'new_records', :new_records, 'new_INCOMPLETE', :new_INCOMPLETE, 'new_closed', coalesce(:new_closed, 0))::VARIANT;
+            -- Figure out the oldest row in the table
+            let range_min timestamp := (select min(end_time) from (
+                select min(end_time) as end_time from internal_reporting_mv.QUERY_HISTORY_COMPLETE_AND_DAILY
```
as above, why hand construction union?

jacques-n commented on bootstrap/019_query_history_updates.sql:
```diff
@@ -67,20 +63,32 @@ BEGIN
         DROP TABLE RAW_QH_EVT;
         COMMIT;
 
+        CALL INTERNAL.SET_CONFIG('QUERY_HISTORY_MAINTENANCE', CURRENT_TIMESTAMP()::string);
```
why move so not always happening? Not sure of intention.

joshelser replied:
I think how we had it before was confusing.

Being outside of the block looked (at first glance) like we would have always called `SET_CONFIG`. However, we were always `RAISE`'ing in the `EXCEPTION` block, so we would have only called `SET_CONFIG` when we didn't catch an exception.

Putting it inside the `BEGIN` block made it more clear to me that we only call this when the block successfully executes.

jacques-n commented on bootstrap/090_post_setup.sql:
```diff
@@ -70,21 +70,84 @@ CREATE OR REPLACE TASK TASKS.WAREHOUSE_EVENTS_MAINTENANCE
+    task_name text default 'SIMPLE_DATA_EVENTS_MAINTENANCE';
+    task_run_id text default (select INTERNAL.TASK_RUN_ID());
+BEGIN
+    let query_id text := (select query_id from table(information_schema.task_history(TASK_NAME => :task_name)) WHERE GRAPH_RUN_GROUP_ID = :task_run_id  AND DATABASE_NAME = current_database() limit 1);
```
i think you can pass in the task run id here. or maybe the task root id. Maybe that would be safer. (Imagine our systems where we have several databases with the same task name.)

jacques-n commented on bootstrap/090_post_setup.sql:
```diff
@@ -70,21 +70,84 @@ CREATE OR REPLACE TASK TASKS.WAREHOUSE_EVENTS_MAINTENANCE
+            INSERT INTO INTERNAL.TASK_LOG(task_start, task_run_id, query_id, input, task_name, object_name) select :start_time, :task_run_id, :query_id, :input, :task_name, :table_name;
+
+            let output variant;
+            CALL INTERNAL.refresh_simple_table(:table_name, :index_col, true, :input) into :output;
```
It seems like the task log handling above should actually be inside the refresh procedure as opposed to in this overall thing.

joshelser replied:
After consolidating the `insert` and `update` into `start_task` and `finish_task`, this feels pretty clean to me and would muddy the waters between the task setup/teardown and task body delineation.

jacques-n commented on bootstrap/090_post_setup.sql:
```diff
@@ -70,21 +70,84 @@ CREATE OR REPLACE TASK TASKS.WAREHOUSE_EVENTS_MAINTENANCE
+BEGIN
+    let query_id text := (select query_id from table(information_schema.task_history(TASK_NAME => :task_name)) WHERE GRAPH_RUN_GROUP_ID = :task_run_id  AND DATABASE_NAME = current_database() limit 1);
+    let input variant := (select output from INTERNAL.TASK_LOG where success AND task_name = :task_name AND object_name = :object_name order by task_start desc limit 1);
+    INSERT INTO INTERNAL.TASK_LOG(task_start, task_run_id, query_id, input, task_name, object_name) select :start_time, :task_run_id, :query_id, :input, :task_name, :object_name;
```
as I look at all these inserts/updates... maybe we should make two procedures that encapsulate these behaviors?

joshelser commented:
I think I've addressed all of your feedback, @jacques-n. Please take another look when you have the chance.

I need to fix up the unit test failure and add some more while I'm in there.

joshelser commented on bootstrap/012_task_log.sql:
```diff
@@ -0,0 +1,119 @@
+$$;
+
+-- Inserts a row into TASK_LOG with the time the task started, the name of the object being materialized, the graph run ID and query_id for the task.
+create or replace procedure internal.start_task(task_name text, object_name text, start_time timestamp_ltz, task_run_id text, query_id text)
```
just noticed that passing a `timestamp_ltz` through a procedure seems to is shifting by the account's default timezone. passing as text doesn't do this.

jacques-n replied:
let's use timestamp_ntz everywhere.

joshelser replied:
ok

rymurr reviewed (approved):
LGTM. I am happy if Jacques is

jacques-n reviewed (commented):
Don't wait for me. If Ryan is happy and you addressed changes let's move forward.

joshelser commented:
Thanks Jacques. No expiration on chance to give more feedback 😆
